controllers/imagerepository_controller.go

```diff
@@ -577,32 +590,216 @@ func (r *ImageRepositoryReconciler) EnsureSecret(ctx context.Context, imageRepos
+	imagePullSecrets := []corev1.LocalObjectReference{}
+	for _, pullSecret := range serviceAccount.ImagePullSecrets {
+		// don't break and search for duplicities
+		if pullSecret.Name == secretName {
+			linksModified = true
+			continue
+		}
+		imagePullSecrets = append(imagePullSecrets, pullSecret)
+	}
+	serviceAccount.ImagePullSecrets = imagePullSecrets
+
+	if linksModified {
+		if err := r.Client.Update(ctx, serviceAccount); err != nil {
+			log.Error(err, "Failed to update pipeline service account", "ServiceAccountName", buildPipelineServiceAccountName, "NamespaceName", namespace, l.Action, l.ActionUpdate)
+			return err
+		}
+		log.Info("Cleaned up secret links in pipeline service account", "SecretName", secretName, "ServiceAccountName", buildPipelineServiceAccountName, l.Action, l.ActionUpdate)
+	}
+	return nil
+}
+
+// VerifyAndFixSecretsLinking ensures that the given secret is linked to the provided service account, and also removes duplicated link for the secret.
+// when secret doesn't exist unlink it from service account
+func (r *ImageRepositoryReconciler) VerifyAndFixSecretsLinking(ctx context.Context, imageRepository *imagerepositoryv1alpha1.ImageRepository) error {
+	log := ctrllog.FromContext(ctx)
+
+	secretName := imageRepository.Status.Credentials.PushSecretName
+
+	// check secret existence and if secret doesn't exist unlink it from service account
+	// users can recreate it by using regenerate-token
+	secret := &corev1.Secret{}
+	secretKey := types.NamespacedName{Namespace: imageRepository.Namespace, Name: secretName}
+	secretExists := true
+	if err := r.Client.Get(ctx, secretKey, secret); err != nil {
+		if !errors.IsNotFound(err) {
+			log.Error(err, "Failed to get secret", "SecretName", secretName, l.Action, l.ActionView)
+			return err
+		}
+
+		secretExists = false
+		log.Info("Secret doesn't exist, will unlink secret from service account", "SecretName", secretName)
+		log.Info("To recreate secret use regenerate-token")
+
+		if err := r.unlinkSecretFromServiceAccount(ctx, secretName, imageRepository.Namespace); err != nil {
+			log.Error(err, "Failed to unlink secret from service account", "SecretName", secretName, l.Action, l.ActionUpdate)
+			return err
 		}
 	}
+
+	if secretExists {
+		// link secret to service account if isn't linked already
+		if err := r.linkSecretToServiceAccount(ctx, secretName, imageRepository.Namespace); err != nil {
+			log.Error(err, "Failed to link secret to service account", "SecretName", secretName, l.Action, l.ActionUpdate)
+			return err
+		}
+
+		// clean duplicate secret links and remove secret from ImagePullSecrets
+		if err := r.cleanUpSecretInServiceAccount(ctx, secretName, imageRepository.Namespace); err != nil {
+			log.Error(err, "Failed to clean up secret in service account", "SecretName", secretName, l.Action, l.ActionUpdate)
+			return err
+		}
+	}
+
+	imageRepository.Spec.Credentials.VerifyLinking = nil
+	if err := r.Client.Update(ctx, imageRepository); err != nil {
```

Comment:
I think it will fail with conflict error because you are trying to update the component many times.
Update the component once or do another reconcile.